pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+		dm.enqueueController(&daemonControllers[i])
+	}
+}
+
+func (dm *DaemonManager) enqueueController(obj interface{}) {
+	key, err := controllerKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	dm.queue.Add(key)
+}
+
+func (dm *DaemonManager) getPodDaemonController(pod *api.Pod) *api.DaemonController {
+	controllers, err := dm.dcStore.GetPodDaemonController(pod)
+	if err != nil {
+		glog.V(4).Infof("No controllers found for pod %v, daemon manager will avoid syncing", pod.Name)
+		return nil
+	}
+	return &controllers[0]
+}
+
+func (dm *DaemonManager) addPod(obj interface{}) {
+	glog.Infoln("Pod added!")
+	pod := obj.(*api.Pod)
+	if dc := dm.getPodDaemonController(pod); dc != nil {
+		dcKey, err := controllerKeyFunc(dc)
+		if err != nil {
+			glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+		}
+		dm.expectations.CreationObserved(dcKey)
+		dm.enqueueController(dc)
+	}
+}
+
+// When a pod is updated, figure out what controller/s manage it and wake them
+// up. If the labels of the pod have changed we need to awaken both the old
+// and new controller. old and cur must be *api.Pod types.
+func (dm *DaemonManager) updatePod(old, cur interface{}) {
+	glog.Infoln("Pod update")
+	if api.Semantic.DeepEqual(old, cur) {
+		// A periodic relist will send update events for all known pods.
+		return
+	}
+	curPod := cur.(*api.Pod)
+	if dc := dm.getPodDaemonController(curPod); dc != nil {
+		dm.enqueueController(dc)
+	}
+	oldPod := old.(*api.Pod)
+	// Only need to get the old controller if the labels changed.
+	if !reflect.DeepEqual(curPod.Labels, oldPod.Labels) {
+		// If the old and new dc are the same, the first one that syncs
+		// will set expectations preventing any damage from the second.
+		if oldRC := dm.getPodDaemonController(oldPod); oldRC != nil {
+			dm.enqueueController(oldRC)
+		}
+	}
+}
+
+func (dm *DaemonManager) deletePod(obj interface{}) {
+	glog.Infoln("Pod deleted!")
+	pod, ok := obj.(*api.Pod)
+
+	// When a delete is dropped, the relist will notice a pod in the store not
+	// in the list, leading to the insertion of a tombstone object which contains
+	// the deleted key/value. Note that this value might be stale. If the pod
+	// changed labels the new rc will not be woken up till the periodic resync.
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*api.Pod)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not a pod %+v", obj)
+			return
+		}
+	}
+	if dc := dm.getPodDaemonController(pod); dc != nil {
+		dcKey, err := controllerKeyFunc(dc)
+		if err != nil {
+			glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+		}
+		dm.expectations.DeletionObserved(dcKey)
+		dm.enqueueController(dc)
+	}
+}
+
+func (dm *DaemonManager) addNode(obj interface{}) {
+	glog.Infoln("Processing Node")
```

Comment:
Add // TODO: it'd be nice to pass a hint with these enqueues, so that each dc would only examine the added node (unless it has other work to do, too).